File: v1/src/models/requests/chamados_request.py
from .base_request import BaseRequest
from models.enums.routes_enum import RoutesEnum
from urllib.parse import urlencode

class ChamadosParamsRequest():
    def __init__(
        self,
        id: int = None,
        situacao: int = None,
        solicitante: int = None,
        tipo: int = None,
        responsavel: int = None,
        unidade: int = None,
        abertoStart: str = None,
        abertoEnd: str = None,
        concluidoStart: str = None,
        concluidoEnd: str = None,
        resolvidoStart: str = None,
        resolvidoEnd: str = None,
        ultimaAlteracaoStart: str = None,
        ultimaAlteracaoEnd: str = None
    ):
        self.start = 0
        self.limit = 100
        self.id = id
        self.situacao = situacao
        self.solicitante = solicitante
        self.tipo = tipo
        self.responsavel = responsavel
        self.unidade = unidade
        self.abertoStart = abertoStart
        self.abertoEnd = abertoEnd
        self.concluidoStart = concluidoStart
        self.concluidoEnd = concluidoEnd
        self.resolvidoStart = resolvidoStart
        self.resolvidoEnd = resolvidoEnd
        self.ultimaAlteracaoStart = ultimaAlteracaoStart
        self.ultimaAlteracaoEnd = ultimaAlteracaoEnd

    def __str__(self):
        return (
            f"Start={self.start}\n"
            f"Limit={self.limit}\n"
            f"Id={self.id or 'Não informado'}\n"
            f"Situacao={self.situacao or 'Não informado'}\n"
            f"Solitante={self.solicitante or 'Não informado'}\n"
            f"Tipo={self.tipo or 'Não informado'}\n"
            f"Responsavel={self.responsavel or 'Não informado'}\n"
            f"Unidade={self.unidade or 'Não informado'}\n"
            f"AbertoStart={self.abertoStart or 'Não informado'}\n"
            f"AbertoEnd={self.abertoEnd or 'Não informado'}\n"
            f"ConcluidoStart={self.concluidoStart or 'Não informado'}\n"
            f"ConcluidoEnd={self.concluidoEnd or 'Não informado'}\n"
            f"ResolvidoStart={self.resolvidoStart or 'Não informado'}\n"
            f"ResolvidoEnd={self.resolvidoEnd or 'Não informado'}\n"
            f"UltimaAlteracaoStart={self.ultimaAlteracaoStart or 'Não informado'}\n"
            f"UltimaAlteracaoEnd={self.ultimaAlteracaoEnd or 'Não informado'}\n"
        )
    
    def getUrlParams(self):
        params = {
            "start": self.start,
            "limit": self.limit,
            "id": self.id,
            "situacao": self.situacao,
            "solicitante": self.solicitante,
            "tipo": self.tipo,
            "responsavel": self.responsavel,
            "unidade": self.unidade,
            "abertoStart": self.abertoStart,
            "abertoEnd": self.abertoEnd,
            "concluidoStart": self.concluidoStart,
            "concluidoEnd": self.concluidoEnd,
            "resolvidoStart": self.resolvidoStart,
            "resolvidoEnd": self.resolvidoEnd,
            "ultimaAlteracaoStart": self.ultimaAlteracaoStart,
            "ultimaAlteracaoEnd": self.ultimaAlteracaoEnd,
        }

        params = {k: v for k, v in params.items() if v is not None}

        return params
    
class ChamadosRequest(BaseRequest):
    def __init__(self, authorization: str, params: ChamadosParamsRequest):
        super().__init__(authorization)
        self.params = params

    def __str__(self):
        return (
            f"Authorization: {self.authorization}\n"
            f"Content-Type: {self.contentType}\n"
            f"{self.params.__str__()}"
        )
    
    def buildUrl(self):
        queryString = urlencode(self.params.getUrlParams())
        return f"{RoutesEnum.Chamados}?{queryString}"
    
    def getHeaders(self):
        return {
            "Authorization": self.authorization,
            "Content-Type": self.contentType
        }

File: v3/src/models/requests/chamados.request.js
const RoutesEnum = require("../enums/routes.enum");

class ChamadosParamsRequest {
  constructor(
    id = null,
    situacao = null,
    solicitante = null,
    tipo = null,
    responsavel = null,
    unidade = null,
    abertoStart = null,
    abertoEnd = null,
    concluidoStart = null,
    concluidoEnd = null,
    resolvidoStart = null,
    resolvidoEnd = null,
    ultimaAlteracaoStart = null,
    ultimaAlteracaoEnd = null
  ) {
    this.start = 0;
    this.limit = 100;
    this.id = id;
    this.situacao = situacao;
    this.solicitante = solicitante;
    this.tipo = tipo;
    this.responsavel = responsavel;
    this.unidade = unidade;
    this.abertoStart = abertoStart;
    this.abertoEnd = abertoEnd;
    this.concluidoStart = concluidoStart;
    this.concluidoEnd = concluidoEnd;
    this.resolvidoStart = resolvidoStart;
    this.resolvidoEnd = resolvidoEnd;
    this.ultimaAlteracaoStart = ultimaAlteracaoStart;
    this.ultimaAlteracaoEnd = ultimaAlteracaoEnd;
  }

  getUrlParams() {
    const params = {
      start: this.start,
      limit: this.limit,
      id: this.id,
      situacao: this.situacao,
      solicitante: this.solicitante,
      tipo: this.tipo,
      responsavel: this.responsavel,
      unidade: this.unidade,
      abertoStart: this.abertoStart,
      abertoEnd: this.abertoEnd,
      concluidoStart: this.concluidoStart,
      concluidoEnd: this.concluidoEnd,
      resolvidoStart: this.resolvidoStart,
      resolvidoEnd: this.resolvidoEnd,
      ultimaAlteracaoStart: this.ultimaAlteracaoStart,
      ultimaAlteracaoEnd: this.ultimaAlteracaoEnd,
    };

    return Object.fromEntries(
      Object.entries(params).filter(([key, value]) => value !== null)
    );
  }
}

class ChamadosRequest {
  constructor(authorization, params) {
    this.authorization = authorization;
    this.params = params;
    this.contentType = "application/json";
  }

  buildUrl() {
    const queryString = new URLSearchParams(
      this.params.getUrlParams()
    ).toString();
    return `${RoutesEnum.Chamados}?${queryString}`;
  }

  getHeaders() {
    return {
      Authorization: this.authorization,
      "Content-Type": this.contentType,
    };
  }
}

module.exports = { ChamadosRequest, ChamadosParamsRequest };
